java-checks/src/main/java/org/sonar/java/checks/IndentationAfterConditionalCheck.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ForEachStatement;
+import org.sonar.plugins.java.api.tree.ForStatementTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.SyntaxToken;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.WhileStatementTree;
+
+@Rule(key = "S3973")
+public class IndentationAfterConditionalCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private JavaFileScannerContext context;
+
+  private Set<IfStatementTree> statementSet = new HashSet<>();
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    this.context = context;
+    statementSet.clear();
+    scan(context.getTree());
```

Comment:
You should clear the set after the scan to, to be sure that references to the `IfStatementTree` are not kept for the whole scanning process.
Actually, I think I'd prefer to avoid using a set altogether by using the parent to retrieve the `else` keyword (i.e. if the parent is a `IfStatementTree` and current if is the else statement of the parent). What do you think?

Reply:
Nice idea!